google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/apis/CloudDependencyWithNoBomInspection.java

```diff
@@ -0,0 +1,170 @@
+
+  @Nullable
+  @Override
+  public String getStaticDescription() {
+    return GctBundle.message("cloud.libraries.with.no.bom.inspection.description");
+  }
+
+  @Override
+  public void checkFileElement(
+      DomFileElement<MavenDomProjectModel> domFileElement, DomElementAnnotationHolder holder) {
+
+    checkDependencyWithNoBom(domFileElement.getRootElement(), holder);
+  }
+
+  private void checkDependencyWithNoBom(
+      MavenDomProjectModel projectModel, DomElementAnnotationHolder holder) {
+    Module module = projectModel.getModule();
+
+    if (module == null) {
+      return;
+    }
+
+    Set<CloudLibrary> cloudLibraries =
+        CloudLibraryProjectState.getInstance(module.getProject()).getCloudLibraries(module);
+
+    if (cloudLibraries.isEmpty()) {
+      return;
+    }
+
+    if (CloudLibraryProjectState.getInstance(module.getProject())
+        .getCloudLibraryBomVersion(module)
+        .isPresent()) {
+      return;
+    }
+
+    projectModel
+        .getDependencies()
+        .getDependencies()
+        .forEach(
+            dependency -> {
+              if (isCloudLibraryDependency(dependency, cloudLibraries)) {
+                holder.createProblem(
+                    dependency,
+                    HighlightSeverity.GENERIC_SERVER_ERROR_OR_WARNING,
+                    GctBundle.message("cloud.libraries.with.no.bom.inspection.problem.description"),
+                    new AddBomAndStripVersionQuickFix(module));
+              }
+            });
+  }
+
+  /**
+   * A {@link LocalQuickFix} that will add the google-cloud-java BOM to the pom.xml and delete the
+   * version tag if specified.
+   */
+  private class AddBomAndStripVersionQuickFix implements LocalQuickFix {
+
+    private Module module;
+
+    AddBomAndStripVersionQuickFix(Module module) {
+      this.module = module;
+    }
+
+    @Nls
+    @NotNull
+    @Override
+    public String getFamilyName() {
+      return GctBundle.message("cloud.libraries.with.no.bom.quickfix.title");
+    }
+
+    @Override
+    public void applyFix(@NotNull Project project, @NotNull ProblemDescriptor descriptor) {
+      XmlTag xmlElement = (XmlTag) descriptor.getPsiElement();
+      if (xmlElement == null) {
+        logger.error(
+            "Unexpected null xml element when attempting to apply DependencyWithNoBom "
+                + "quick-fix");
+        return;
+      }
+
+      addBom(project);
+
+      // Strip out the version tag if present
+      Stream.of(xmlElement.getChildren())
+          .filter(childTag -> childTag instanceof XmlTag)
+          .map(childTag -> (XmlTag) childTag)
+          .filter(childTag -> "version".equals(childTag.getName()))
+          .findFirst()
+          .ifPresent(CloudDependencyWithNoBomInspection.this::stripVersion);
+    }
+
+    private void addBom(Project project) {
+      MavenProject mavenProject = MavenProjectsManager.getInstance(project).findProject(module);
```

Comment:
Just to clarify a bit, `MavenProject` maps to `Module`, and there could be multiple `MavenProject`s per IDE `Project`, hence the need for the module field, correct?

Reply:
correct. this inspection in the scope of a particular pom.xml which is part of a particular module.